client/context/query.go

```diff
@@ -0,0 +1,309 @@
+	}
+
+	if len(accountBytes) == 0 {
+		return ErrInvalidAccount(addr)
+	}
+
+	return nil
+}
+
+// EnsureBroadcastTx broadcasts a transactions either synchronously or
+// asynchronously based on the context parameters. The result of the broadcast
+// is parsed into an intermediate structure which is logged if the context has
+// a logger defined.
+func (ctx QueryContext) EnsureBroadcastTx(txBytes []byte) error {
+	if ctx.Async {
+		return ctx.ensureBroadcastTxAsync(txBytes)
+	}
+
+	return ctx.ensureBroadcastTx(txBytes)
+}
+
+func (ctx QueryContext) ensureBroadcastTxAsync(txBytes []byte) error {
+	res, err := ctx.BroadcastTxAsync(txBytes)
+	if err != nil {
+		return err
+	}
+
+	if ctx.JSON {
+		type toJSON struct {
+			TxHash string
+		}
+
+		if ctx.Logger != nil {
+			resJSON := toJSON{res.Hash.String()}
+			bz, err := ctx.Codec.MarshalJSON(resJSON)
+			if err != nil {
+				return err
+			}
+
+			ctx.Logger.Write(bz)
+			io.WriteString(ctx.Logger, "\n")
+		}
+	} else {
+		if ctx.Logger != nil {
+			io.WriteString(ctx.Logger, fmt.Sprintf("Async tx sent (tx hash: %s)\n", res.Hash))
+		}
+	}
+
+	return nil
+}
+
+func (ctx QueryContext) ensureBroadcastTx(txBytes []byte) error {
+	res, err := ctx.BroadcastTx(txBytes)
+	if err != nil {
+		return err
+	}
+
+	if ctx.JSON {
+		// since JSON is intended for automated scripts, always include
+		// response in JSON mode.
+		type toJSON struct {
+			Height   int64
+			TxHash   string
+			Response string
+		}
+
+		if ctx.Logger != nil {
+			resJSON := toJSON{res.Height, res.Hash.String(), fmt.Sprintf("%+v", res.DeliverTx)}
+			bz, err := ctx.Codec.MarshalJSON(resJSON)
+			if err != nil {
+				return err
+			}
+
+			ctx.Logger.Write(bz)
+			io.WriteString(ctx.Logger, "\n")
+		}
+
+		return nil
+	}
+
+	if ctx.Logger != nil {
+		resStr := fmt.Sprintf("Committed at block %d (tx hash: %s)\n", res.Height, res.Hash.String())
+
+		if ctx.PrintResponse {
+			resStr = fmt.Sprintf("Committed at block %d (tx hash: %s, response: %+v)\n",
+				res.Height, res.Hash.String(), res.DeliverTx,
+			)
+		}
+
+		io.WriteString(ctx.Logger, resStr)
+	}
+
+	return nil
+}
+
+// query performs a query from a Tendermint node with the provided store name
+// and path.
+func (ctx QueryContext) query(path string, key common.HexBytes) (res []byte, err error) {
+	node, err := ctx.GetNode()
+	if err != nil {
+		return res, err
+	}
+
+	opts := rpcclient.ABCIQueryOptions{
+		Height:  ctx.Height,
+		Trusted: ctx.TrustNode,
+	}
+
+	result, err := node.ABCIQueryWithOptions(path, key, opts)
+	if err != nil {
+		return res, err
+	}
+
+	resp := result.Response
+	if !resp.IsOK() {
+		return res, errors.Errorf("query failed: (%d) %s", resp.Code, resp.Log)
+	}
+
+	return resp.Value, nil
+}
+
+// queryStore performs a query from a Tendermint node with the provided a store
+// name and path.
+func (ctx QueryContext) queryStore(key cmn.HexBytes, storeName, endPath string) ([]byte, error) {
+	path := fmt.Sprintf("/store/%s/%s", storeName, endPath)
```

Comment:
maybe "store" here should likewise be parameterized

Reply:
That prefix (`/store` for store queries) is hardcoded in baseapp